clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -102,6 +102,7 @@ static ResourceClass getResourceClass(RegisterType RT) {
     return ResourceClass::Sampler;
   case RegisterType::C:
   case RegisterType::I:
+  default:
     llvm_unreachable("unexpected RegisterType value");
   }
```

Comment:
There is a very useful warning about switching on an enumerator where not all enumerators are covered.

Adding a default case suppresses this warning, but this is undesirable, as we want to see this warning pop up when we add a new enumerator.

```suggestion
  }
  llvm_unreachable("unexpected RegisterType value");
```

Reply:
I don't agree with the suggested change of not using default.

@hekota  says:
> the switch covers all enum values. 

It sounds like  line 107 will be impossible to hit.

So then whats the difference between an unexpected and an unhandled register type? Why couldn't we have just done a default case?  Why can't MSVC2022 determine that all cases are covered? Are we really making a change to appease a false positive from a compiler warning?

If a new Register Type were added it would hit a llvm_unreachable on the default case and an implementer would quickly know to implement it from the call stack. I don't like the idea of changes to make a warning happy when that code change  adds undistinguishable terminology between unexpected and  unhandled. I would prefer not adding new terminology that would be confusing and instead add a default.

Reply:
> It sounds like line 107 will be impossible to hit.

Not impossible: someone can always cast an arbitrary value into the enum type, which is what MSVC is trying to guard against.

As a compromise we could keep the unreachable at the bottom, and then deliberately fall through to that in the unhandled cases:
```cpp
  case RegisterType::C:
  case RegisterType::I:
    // Deliberately falling through to the unreachable below.
    break;
  }
  llvm_unreachable("unexpected RegisterType value");
}
```

Reply:
> As a compromise we could keep the unreachable at the bottom, and then deliberately fall through to that in the unhandled cases:

FWIW, that's the usual approach we take.

Reply:
I'm fine with that.